reporters/ReportVectorMigration.cpp

```diff
@@ -166,33 +220,61 @@ GetReportInstantiator( Kernel::instantiator_function_t* pif )
             throw QueryInterfaceException(__FILE__, __LINE__, __FUNCTION__, "pvc", "IMigrate", "IVectorCohort");
         }
 
+        VectorStateEnum::Enum state = pvc->GetState();
+        if (!m_MustBeInState.empty())
+        {
+            if (std::find(m_MustBeInState.begin(), m_MustBeInState.end(), state) == m_MustBeInState.end())
+            {
+                // state is not one of the required states
+                return;
+            }
+        }
+
         uint32_t vci_id = pvc->GetID();
         int from_node_id = pSim->GetNodeExternalID( nodeSuid ) ;
+        if (!m_MustBeFromNode.empty())
+        {
+            if (std::find(m_MustBeFromNode.begin(), m_MustBeFromNode.end(), from_node_id) == m_MustBeFromNode.end())
+            {
+                // from_node_id is not one we want in the report
+                return;
+            }
+        }
+
         int to_node_id = pSim->GetNodeExternalID( pim->GetMigrationDestination() ) ;
-        int mig_type = pim->GetMigrationType() ;
-        VectorStateEnum::Enum state = pvc->GetState();
+        if (!m_MustBeToNode.empty())
+        {
+            if (std::find(m_MustBeToNode.begin(), m_MustBeToNode.end(), to_node_id) == m_MustBeToNode.end())
+            {
+                // to_node_id is not one we want in the report
+                return;
+            }
+        }
         std::string species = pvc->GetSpecies();
+        if (!m_SpeciesList.empty())
+        {
+            if (std::find(m_SpeciesList.begin(), m_SpeciesList.end(), species) == m_SpeciesList.end())
+            {
+                // if not species we're interested in do not add to report
+                return;
+            }
+        }
```

Comment:
Above you have the same code repeated four times.  What about replacing it with something like the following:

```
    template<typename T>
    bool Qualifies( const std::vector<T>& rList, const T& rValue )
    {
        if( rList.empty() )
        {
            return true;
        }
        else
        {
            return std::find( rList.begin(), rList.end(), rValue ) != rList.end();
        }
    }

    VectorStateEnum::Enum state        =  pvc->GetState();
    int                   from_node_id =  pSim->GetNodeExternalID( nodeSuid ) ;
    int                   to_node_id   =  pSim->GetNodeExternalID( pim->GetMigrationDestination() ) ;
    const std::string    species       =  pvc->GetSpecies(); 

    if( !Qualifies<ExternalNodeId_t     >( m_MustBeFromNode, from_node_id  ) ) return;
    if( !Qualifies<ExternalNodeId_t     >( m_MustBeToNode,   to_node_id    ) ) return;
    if( !Qualifies<VectorStateEnum::Enum>( m_MustBeInState,  state         ) ) return;
    if( !Qualifies<std::string          >( m_SpeciesList,    species       ) ) return;
```